pkg/component/controller/leaderelector.go

```diff
@@ -47,7 +46,10 @@ type leaderElector struct {
 }
 
 // NewLeaderElector creates new leader elector
-func NewLeaderElector(kubeClientFactory kubeutil.ClientFactoryInterface) LeaderElector {
+func NewLeaderElector(kubeClientFactory kubeutil.ClientFactoryInterface) interface {
+	LeaderElector
+	component.Component
```

Comment:
What is the idea here?

Reply:
Separation of concerns. We have two use cases for the return value: Once, for the component manager (which doesn't care about the component being a leader elector), and once for the rest of the code, which doesn't care that the leader elector is a component, and shouldn't call any of its lifecycle methods.

Reply:
Yeah, maybe we can do it this way:

```
type LeaderElectorComponent interface {
	LeaderElector
	component.Component
}

func NewLeaderElector(kubeClientFactory kubeutil.ClientFactoryInterface) LeaderElectorComponent 
```

Reply:
Ah okay, so the answer here was "laziness" ;) I thought we could go without the extra type. Would you prefer this?

Reply:
I don't know the reasons why we don't return just struct here, but yeah, if we decide to return an interface, I think having something like `LeaderElectorComponent` is a bit clearer.

Reply:
:-/

* Lint: pkg/component/controller/leaderelector.go#L52
  unexported-return: exported func NewLeaderElector returns unexported type *controller.leaderElector, which can be annoying to use (revive)
* Lint: pkg/component/controller/k0scloudprovider.go#L40
  unexported-return: exported func NewK0sCloudProvider returns unexported type *controller.k0sCloudProvider, which can be annoying to use (revive)

I am inclined to either silence the linter in those cases via a nolint comment, or return interfaces instead :disappointed: WDYT @makhov?

Reply:
I would definitely vote for named interface. You write it once and read X times and each time reading the anonymous iface would require from me unnecessary concentration.

Reply:
For the "silence the linter" case, there would be only a single occurrence of a composite interface left. Not sure if it would make sense to declare a type alias just for that.

For the "return interface" case, there would still be just two occurrences of the composite interface. That's at least one more, so that type alias would make some sense.